app/waf/src/repositories/common.ts

```diff
@@ -15,11 +24,80 @@ export class CommonRepository<
     return super.create(data, options);
   }
 
-  updateById(id: ID, data: DataObject<T>, options?: Options): Promise<void> {
-    Object.assign(data, {
-      updatedAt: new Date().toISOString(),
-    });
+  find(filter?: Filter<T>, options?: Options, tenantId?: string): Promise<T[]> {
+    if (tenantId) {
+      let tenantFilter: Filter = {where: {tenantId: tenantId}};
+      return super.find(tenantFilter as (Filter<T>), options);
+    } else {
+      return super.find(filter, options);
+    }
+  }
+
+  async findById(
+    id: ID,
+    filter?: Filter<T>,
+    options?: Options,
+    tenantId?: string,
```

Comment:
Why not move tenantId to the second position?

Reply:
```
Property 'find' in type 'CommonRepository<T, ID>' is not assignable to the same property in base type 'DefaultCrudRepository<T, ID>'.
  Type '(filter?: Filter<T> | undefined, tenantId?: string | undefined, options?: AnyObject | undefined) => Promise<T[]>' is not assignable to type '(filter?: Filter<T> | undefined, options?: AnyObject | undefined) => Promise<T[]>'.
    Types of parameters 'tenantId' and 'options' are incompatible.
      Type 'AnyObject | undefined' is not assignable to type 'string | undefined'.
        Type 'AnyObject' is not assignable to type 'string'.ts(2416)
```

it will report this. the same function name with base class, but the optional argument is not the same.

Reply:
unless I make the tenantId as Options type too.

Reply:
find(filter?: Filter<T>, tenantId?: Options, options?: Options, ): Promise<T[]>

Reply:
@zhaoqin-github 
   ```
    let f: Filter = filter || {};
    let builder = new WhereBuilder(f.where);
    f.where = builder.and({id: 'abcd'}).build();
  ```

I test the way you suggested by the code below:
```
Control layer
  @get(prefix + '/monitors', {
    responses: {
      '200': Schema.collectionResponse(
        Monitor,
        'Successfully retrieve Monitor resources',
      ),
    },
  })
  async find(
    @param.query.object('filter', getFilterSchemaFor(Monitor)) filter?: Filter,
  ): Promise<CollectionResponse> {
    let tenantId = await this.reqCxt.get(WafBindingKeys.Request.KeyTenantId);
    let data = await this.monitorRepository.find(
      undefined,
      undefined,
      tenantId,
    );
    return new CollectionResponse(Monitor, data);
  }
----------------------------------------------------------------------------
Repository layer:

  find(filter?: Filter<T>, options?: Options, tenantId?: string): Promise<T[]> {
    if (tenantId) {
      let f: Filter = filter || {};
      let builder = new WhereBuilder(f.where);
      f.where = builder.and({ where: { tenantId: tenantId } }).build();
      console.log("*********************");
      console.log(f);
      console.log("*********************");
      return super.find(f as (Filter<T>), options);
    } else {
      return super.find(filter, options);
    }
  }
----------------------------------------------------------------------------
Call with REST API:
http://127.0.0.1:3000/adcaas/v1/monitors
----------------------------------------------------------------------------
Get response:
{
    "monitors": [
        {
            "id": "00379ad5-0f57-4155-9d7a-b43244bd9b91",
            "name": "a1",
            "description": "My description",
            "tenantId": "fdac59f5b20046829ea58720702axxxx",
            "createdAt": "2019-04-28T02:39:52.759Z",
            "updatedAt": null,
            "interval": 5,
            "timeout": 16,
            "targetAddress": "192.168.10.123",
            "targetPort": 8080,
            "monitorType": "tcp"
        },
        {
            "id": "12321313",
            "name": "My name",
            "description": "My description",
            "tenantId": "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
            "createdAt": null,
            "updatedAt": null,
            "interval": 5,
            "timeout": 16,
            "targetAddress": "192.168.10.123",
            "targetPort": 8080,
            "monitorType": "tcp"
        },
        {
            "id": "21043181-67ba-44d2-addc-fd8ae1bc0c30",
            "name": "a4",
            "description": "My description",
            "tenantId": "fdac59f5b20046829ea58720702axxxx",
            "createdAt": "2019-04-28T02:40:12.898Z",
            "updatedAt": null,
            "interval": 5,
            "timeout": 16,
            "targetAddress": "192.168.10.123",
            "targetPort": 8080,
            "monitorType": "tcp"
        },
        {
            "id": "54fb6cfc-29b2-473b-b61d-54e0838596bf",
            "name": "a2",
            "description": "My description",
            "tenantId": "fdac59f5b20046829ea58720702axxxx",
            "createdAt": "2019-04-28T02:39:57.700Z",
            "updatedAt": null,
            "interval": 5,
            "timeout": 16,
            "targetAddress": "192.168.10.123",
            "targetPort": 8080,
            "monitorType": "tcp"
        },
        {
            "id": "57381921-fa9c-4ce3-9377-92de795b7603",
            "name": "a3",
            "description": "My description",
            "tenantId": "fdac59f5b20046829ea58720702axxxx",
            "createdAt": "2019-04-28T02:40:07.191Z",
            "updatedAt": null,
            "interval": 5,
            "timeout": 16,
            "targetAddress": "192.168.10.123",
            "targetPort": 8080,
            "monitorType": "tcp"
        },
        {
            "id": "854aaf84-3a3f-4b50-8558-ba3bf05e84fb",
            "name": "a2",
            "description": "My description",
            "tenantId": "fdac59f5b20046829ea58720702axxxx",
            "createdAt": "2019-04-28T02:40:03.271Z",
            "updatedAt": null,
            "interval": 5,
            "timeout": 16,
            "targetAddress": "192.168.10.123",
            "targetPort": 8080,
            "monitorType": "tcp"
        },
        {
            "id": "a5582db7-e22b-4103-a5e0-08884d6d3d9d",
            "name": "My name",
            "description": "My description",
            "tenantId": "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
            "createdAt": "2019-04-02T07:55:19.538Z",
            "updatedAt": null,
            "interval": 5,
            "timeout": 16,
            "targetAddress": "192.168.10.123",
            "targetPort": 8080,
            "monitorType": "tcp"
        },
        {
            "id": "dcc68f15-8d96-4882-ae38-3bf3f35bcc56",
            "name": "t4",
            "description": "My description",
            "tenantId": "fdac59f5b20046829ea58720702a74af",
            "createdAt": "2019-04-28T02:39:33.814Z",
            "updatedAt": null,
            "interval": 5,
            "timeout": 16,
            "targetAddress": "192.168.10.123",
            "targetPort": 8080,
            "monitorType": "tcp"
        },
        {
            "id": "ee6b88c3-c5fd-4a5a-9fe4-a6df05471203",
            "name": "string",
            "description": null,
            "tenantId": null,
            "createdAt": null,
            "updatedAt": null,
            "interval": 0,
            "timeout": 0,
            "targetAddress": "string",
            "targetPort": 0,
            "monitorType": "string"
        },
        {
            "id": "fa1f961a-6681-4319-94d8-6478472bc447",
            "name": "t3",
            "description": "My description",
            "tenantId": "fdac59f5b20046829ea58720702a74af",
            "createdAt": "2019-04-28T02:39:28.477Z",
            "updatedAt": null,
            "interval": 5,
            "timeout": 16,
            "targetAddress": "192.168.10.123",
            "targetPort": 8080,
            "monitorType": "tcp"
        }
    ]
}
----------------------------------------------------------------------------
LOG:
2019-04-30 15:04:24,929 INFO [api.call] Request: {"uuid":"2fd74380-6b16-11e9-a268-0793695baa3b","method":"GET","headers":{"x-auth-token":"gAAAAABcx_MCjznpYOicQHtC1Z34XfVtH0EkmM5OV5nHDPF7IpJ367IPtc70XB2N-gVjgeWF2oaPZ7u-j59diYAqQEaw5HQd3G2dE1Sqws21sDtnkUnwIZX5ah8rvyeo5XOaAFeAIwBu4KjQVUywmyupFH2-imIqEXdKfA_Mj3XNZg6fdP7PD10","tenant-id":"fdac59f5b20046829ea58720702a74af","content-type":"application/json","user-agent":"PostmanRuntime/7.11.0","accept":"*/*","cache-control":"no-cache","postman-token":"f8be5d68-cc45-4fff-9d1f-e42a321ee391","host":"127.0.0.1:3000","accept-encoding":"gzip, deflate","content-length":"84","connection":"keep-alive"},"path":"/adcaas/v1/monitors","params":{}}
2019-04-30 15:04:24,930 DEBUG [api.call] start to authenticate user
(node:12465) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 connected listeners added. Use emitter.setMaxListeners() to increase limit
2019-04-30 15:04:26,122 DEBUG [api.call] Authenticated OK
2019-04-30 15:04:26,123 DEBUG [api.call] {"region":"RegionOne","interface":"internal","version":"v3","issuedAt":"2019-04-30T07:02:26.000Z","expiredAt":"2019-04-30T08:02:26.000Z","token":"gAAAAABcx_MCjznpYOicQHtC1Z34XfVtH0EkmM5OV5nHDPF7IpJ367IPtc70XB2N-gVjgeWF2oaPZ7u-j59diYAqQEaw5HQd3G2dE1Sqws21sDtnkUnwIZX5ah8rvyeo5XOaAFeAIwBu4KjQVUywmyupFH2-imIqEXdKfA_Mj3XNZg6fdP7PD10","userId":"934d8da3f75d4bf3b07375a8701f29ca","catalog":[{"endpoints":[{"url":"http://10.145.65.10/placement","interface":"internal","region":"RegionOne","region_id":"RegionOne","id":"0f75bfa5baf64ddca7a30dd4bd0f7af8"},{"url":"http://10.145.65.10/placement","interface":"public","region":"RegionOne","region_id":"RegionOne","id":"34a83f5d7e094965988d3c36311980f6"},{"url":"http://10.145.65.10/placement","interface":"admin","region":"RegionOne","region_id":"RegionOne","id":"ab5d31ddc96c47fbba1d97c836812c6e"}],"type":"placement","id":"1f15566d94c742938bc3d8ef8cdc74e5","name":"placement"},{"endpoints":[{"url":"http://10.145.65.10:8774/v2.1","interface":"internal","region":"RegionOne","region_id":"RegionOne","id":"6b663a069cab4d8d8626a87e5a9cfceb"},{"url":"http://10.145.65.10:8774/v2.1","interface":"admin","region":"RegionOne","region_id":"RegionOne","id":"963eca6fa7854da299ab443a6fc5e1cd"},{"url":"http://10.145.65.10:8774/v2.1","interface":"public","region":"RegionOne","region_id":"RegionOne","id":"99ff2ca2b0b74bfb96762d578bd6d75b"}],"type":"compute","id":"38c02cac0cc842de9cbf2ca97fd41131","name":"nova"},{"endpoints":[{"url":"http://10.145.65.10:9696/","interface":"admin","region":"RegionOne","region_id":"RegionOne","id":"5cfb9cb93cac42d49b62268184d20fc1"},{"url":"http://10.145.65.10:9696/","interface":"internal","region":"RegionOne","region_id":"RegionOne","id":"9297437c4c93492bb9edc7df3a7c8436"},{"url":"http://10.145.65.10:9696/","interface":"public","region":"RegionOne","region_id":"RegionOne","id":"ca1791b2656440819be7bb8c0857fd80"}],"type":"network","id":"8a52dd11d0b14fc1bac7444b7fb6435f","name":"neutron"},{"endpoints":[{"url":"http://10.145.65.10/identity","interface":"public","region":"RegionOne","region_id":"RegionOne","id":"9bf7680f124747088dff9447d7144ee2"},{"url":"http://10.145.65.10/identity","interface":"internal","region":"RegionOne","region_id":"RegionOne","id":"bd0c049702eb4c1687def04167c793b9"},{"url":"http://10.145.65.10/identity_admin","interface":"admin","region":"RegionOne","region_id":"RegionOne","id":"ebca889ec59143abb5f79a066b6f55a7"}],"type":"identity","id":"8c42cb8b0b0f4c9aac185099b0318701","name":"keystone"},{"endpoints":[{"url":"http://10.145.65.10:8774/v2/fdac59f5b20046829ea58720702a74af","interface":"public","region":"RegionOne","region_id":"RegionOne","id":"7a8afadfa68f4e078e055bca8b0da07f"},{"url":"http://10.145.65.10:8774/v2/fdac59f5b20046829ea58720702a74af","interface":"internal","region":"RegionOne","region_id":"RegionOne","id":"a6b035df78b04deba63c8b1684d9ee54"},{"url":"http://10.145.65.10:8774/v2/fdac59f5b20046829ea58720702a74af","interface":"admin","region":"RegionOne","region_id":"RegionOne","id":"efd2dabe1fde4632964f346e35934006"}],"type":"compute_legacy","id":"f2c9f265b1c54f27a94a913cb155f355","name":"nova_legacy"},{"endpoints":[{"url":"http://10.145.65.10:9292","interface":"admin","region":"RegionOne","region_id":"RegionOne","id":"1be1fad26f4e4ffca0bf668a5909f203"},{"url":"http://10.145.65.10:9292","interface":"public","region":"RegionOne","region_id":"RegionOne","id":"5d8c3f7b501147b894b939e5898fa98d"},{"url":"http://10.145.65.10:9292","interface":"internal","region":"RegionOne","region_id":"RegionOne","id":"8b5d0e482e9f44f6aab104c611b59eaf"}],"type":"image","id":"feff310a71b641bf83e069a7885d75c9","name":"glance"}],"tenantId":"fdac59f5b20046829ea58720702a74af"}
*********************
{ where: { and: [ [Object] ] } }
*********************
2019-04-30 15:04:26,142 INFO [api.call] Response: {"uuid":"2fd74380-6b16-11e9-a268-0793695baa3b","statusCode":200,"statusMessage":"OK","headers":{"x-powered-by":"Express","access-control-allow-origin":"*","access-control-allow-credentials":"true","content-type":"application/json"},"body":{"monitors":[{"id":"00379ad5-0f57-4155-9d7a-b43244bd9b91","name":"a1","description":"My description","tenantId":"fdac59f5b20046829ea58720702axxxx","createdAt":"2019-04-28T02:39:52.759Z","updatedAt":null,"interval":5,"timeout":16,"targetAddress":"192.168.10.123","targetPort":8080,"monitorType":"tcp"},{"id":"12321313","name":"My name","description":"My description","tenantId":"aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee","createdAt":null,"updatedAt":null,"interval":5,"timeout":16,"targetAddress":"192.168.10.123","targetPort":8080,"monitorType":"tcp"},{"id":"21043181-67ba-44d2-addc-fd8ae1bc0c30","name":"a4","description":"My description","tenantId":"fdac59f5b20046829ea58720702axxxx","createdAt":"2019-04-28T02:40:12.898Z","updatedAt":null,"interval":5,"timeout":16,"targetAddress":"192.168.10.123","targetPort":8080,"monitorType":"tcp"},{"id":"54fb6cfc-29b2-473b-b61d-54e0838596bf","name":"a2","description":"My description","tenantId":"fdac59f5b20046829ea58720702axxxx","createdAt":"2019-04-28T02:39:57.700Z","updatedAt":null,"interval":5,"timeout":16,"targetAddress":"192.168.10.123","targetPort":8080,"monitorType":"tcp"},{"id":"57381921-fa9c-4ce3-9377-92de795b7603","name":"a3","description":"My description","tenantId":"fdac59f5b20046829ea58720702axxxx","createdAt":"2019-04-28T02:40:07.191Z","updatedAt":null,"interval":5,"timeout":16,"targetAddress":"192.168.10.123","targetPort":8080,"monitorType":"tcp"},{"id":"854aaf84-3a3f-4b50-8558-ba3bf05e84fb","name":"a2","description":"My description","tenantId":"fdac59f5b20046829ea58720702axxxx","createdAt":"2019-04-28T02:40:03.271Z","updatedAt":null,"interval":5,"timeout":16,"targetAddress":"192.168.10.123","targetPort":8080,"monitorType":"tcp"},{"id":"a5582db7-e22b-4103-a5e0-08884d6d3d9d","name":"My name","description":"My description","tenantId":"aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee","createdAt":"2019-04-02T07:55:19.538Z","updatedAt":null,"interval":5,"timeout":16,"targetAddress":"192.168.10.123","targetPort":8080,"monitorType":"tcp"},{"id":"dcc68f15-8d96-4882-ae38-3bf3f35bcc56","name":"t4","description":"My description","tenantId":"fdac59f5b20046829ea58720702a74af","createdAt":"2019-04-28T02:39:33.814Z","updatedAt":null,"interval":5,"timeout":16,"targetAddress":"192.168.10.123","targetPort":8080,"monitorType":"tcp"},{"id":"ee6b88c3-c5fd-4a5a-9fe4-a6df05471203","name":"string","description":null,"tenantId":null,"createdAt":null,"updatedAt":null,"interval":0,"timeout":0,"targetAddress":"string","targetPort":0,"monitorType":"string"},{"id":"fa1f961a-6681-4319-94d8-6478472bc447","name":"t3","description":"My description","tenantId":"fdac59f5b20046829ea58720702a74af","createdAt":"2019-04-28T02:39:28.477Z","updatedAt":null,"interval":5,"timeout":16,"targetAddress":"192.168.10.123","targetPort":8080,"monitorType":"tcp"}]}}
```

This code seems not work.

Reply:
Let me withdraw this comment. Typescript does not support that.